### PR TITLE
Onefact 12403 12404

### PR DESCRIPTION
**Pull request by travislondon**
[Issues: #12403 #12404](https://github.com/xtuml/bridgepoint/commit/a286b0210822b4b0b75dae11cece6ef1fb0fe40b) 

Multiple bug fixes applied:

1. Marking information is cached in a map on workspace startup
   for each project.  Nothing was done to handle the case where
   a project was closed and reopened, or deleted.  This can lead
   to data synchronization issues.  A resource listener is added
   to remove the entry for a closed project key, and recreate from
   disk on open.  The same removal is done for project deletions.

2. The Mark element holds a reference to the marked
   NonRootModelElement.  On model reloads, through things like
   configuration management, this reference can become
   disconnected from its parent.  This causes the MarkingData
   logic to consider this a removal as the element's path is only
   the element name.  To address this access to the NRME reference
   is changed such that it must go through a get operation.  This
   operation is designed to dynamically retrieve the current NRME
   via the hashmap lookup support found in the ModelRoot.

3. Creation of feature.marking after project creation is not
   recognized by the already populated feature map.  The cost of
   attempting to populate the features (which looks for and reads
   the feature.marking file) is small.  This call is added when
   opening the Marking Dialog if the feature map is empty.

4. Creation of application.marking for the first time is
   unsuccessful as the call to write does not account for the file
   not being present.  The only workaround would be to create the
   file manually.  This is addressed by checking for existence, if
   it exists set the file contents as original.  If it does not
   exist use the file.create() function.

### CODE REVIEW

File: src/org.xtuml.bp.ui.marking/src/org/xtuml/bp/ui/marking/MarkingData.java [hunk 1]
```diff
@@ -149,6 +161,11 @@ private void populateMarkings() {
 	 * @return String - An error message if issues were found, otherwise empty string
 	 */
 	public String validateFeatures() {
+		// doesn't hurt to populate here, incase the features file was added
+		// after the project creation
+		if(featureMap.isEmpty()) {
+			populateFeatures();
+		}
```
Comment (leviathan747) [inline src/org.xtuml.bp.ui.marking/src/org/xtuml/bp/ui/marking/MarkingData.java:168]: What would happen if I edit "features.mark" (e.g. via config management) and refresh the project (assuming I already had some features loaded in the map)? Do I need to close and re-open BridgePoint?

**Review by leviathan747 — CHANGES_REQUESTED** [review]
Please add an implementation note so the documentation goes into the repository.

File: src/org.xtuml.bp.ui.marking/src/org/xtuml/bp/ui/marking/MarkingData.java (hunk 1 above)
Comment (travislondon) [reply src/org.xtuml.bp.ui.marking/src/org/xtuml/bp/ui/marking/MarkingData.java:168]: Right now if you were to make changes to the features.mark file and refresh nothing would happen.  You would not get those new changes and your would not lose the in-memory changes.  You would lose in-memory changes if you were to close the project and reopen it.  This issue does not attempt to address a marking file listener and merge in-memory changes with those on disk.  Another issue should be raised if such an extension of the marking functionality is desired.

**Review by leviathan747 — APPROVED** [review]
Pending merge of my PR into your branch, this will be ready for merge

**Pull request merged**